// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: resources/sync/data.proto

package sync

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on DataStatus with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DataStatus) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DataStatus with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DataStatusMultiError, or
// nil if none found.
func (m *DataStatus) ValidateAll() error {
	return m.validate(true)
}

func (m *DataStatus) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Count

	if len(errors) > 0 {
		return DataStatusMultiError(errors)
	}

	return nil
}

// DataStatusMultiError is an error wrapping multiple validation errors
// returned by DataStatus.ValidateAll() if the designated constraints aren't met.
type DataStatusMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DataStatusMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DataStatusMultiError) AllErrors() []error { return m }

// DataStatusValidationError is the validation error returned by
// DataStatus.Validate if the designated constraints aren't met.
type DataStatusValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataStatusValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataStatusValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataStatusValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataStatusValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataStatusValidationError) ErrorName() string { return "DataStatusValidationError" }

// Error satisfies the builtin error interface
func (e DataStatusValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDataStatus.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataStatusValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataStatusValidationError{}

// Validate checks the field values on DataJobs with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DataJobs) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DataJobs with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DataJobsMultiError, or nil
// if none found.
func (m *DataJobs) ValidateAll() error {
	return m.validate(true)
}

func (m *DataJobs) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetJobs()) > 200 {
		err := DataJobsValidationError{
			field:  "Jobs",
			reason: "value must contain no more than 200 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	for idx, item := range m.GetJobs() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DataJobsValidationError{
						field:  fmt.Sprintf("Jobs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DataJobsValidationError{
						field:  fmt.Sprintf("Jobs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DataJobsValidationError{
					field:  fmt.Sprintf("Jobs[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return DataJobsMultiError(errors)
	}

	return nil
}

// DataJobsMultiError is an error wrapping multiple validation errors returned
// by DataJobs.ValidateAll() if the designated constraints aren't met.
type DataJobsMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DataJobsMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DataJobsMultiError) AllErrors() []error { return m }

// DataJobsValidationError is the validation error returned by
// DataJobs.Validate if the designated constraints aren't met.
type DataJobsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataJobsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataJobsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataJobsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataJobsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataJobsValidationError) ErrorName() string { return "DataJobsValidationError" }

// Error satisfies the builtin error interface
func (e DataJobsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDataJobs.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataJobsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataJobsValidationError{}

// Validate checks the field values on DataUsers with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DataUsers) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DataUsers with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DataUsersMultiError, or nil
// if none found.
func (m *DataUsers) ValidateAll() error {
	return m.validate(true)
}

func (m *DataUsers) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetUsers()) > 500 {
		err := DataUsersValidationError{
			field:  "Users",
			reason: "value must contain no more than 500 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	for idx, item := range m.GetUsers() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DataUsersValidationError{
						field:  fmt.Sprintf("Users[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DataUsersValidationError{
						field:  fmt.Sprintf("Users[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DataUsersValidationError{
					field:  fmt.Sprintf("Users[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return DataUsersMultiError(errors)
	}

	return nil
}

// DataUsersMultiError is an error wrapping multiple validation errors returned
// by DataUsers.ValidateAll() if the designated constraints aren't met.
type DataUsersMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DataUsersMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DataUsersMultiError) AllErrors() []error { return m }

// DataUsersValidationError is the validation error returned by
// DataUsers.Validate if the designated constraints aren't met.
type DataUsersValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataUsersValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataUsersValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataUsersValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataUsersValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataUsersValidationError) ErrorName() string { return "DataUsersValidationError" }

// Error satisfies the builtin error interface
func (e DataUsersValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDataUsers.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataUsersValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataUsersValidationError{}

// Validate checks the field values on DataVehicles with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DataVehicles) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DataVehicles with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DataVehiclesMultiError, or
// nil if none found.
func (m *DataVehicles) ValidateAll() error {
	return m.validate(true)
}

func (m *DataVehicles) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetVehicles()) > 1000 {
		err := DataVehiclesValidationError{
			field:  "Vehicles",
			reason: "value must contain no more than 1000 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	for idx, item := range m.GetVehicles() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DataVehiclesValidationError{
						field:  fmt.Sprintf("Vehicles[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DataVehiclesValidationError{
						field:  fmt.Sprintf("Vehicles[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DataVehiclesValidationError{
					field:  fmt.Sprintf("Vehicles[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return DataVehiclesMultiError(errors)
	}

	return nil
}

// DataVehiclesMultiError is an error wrapping multiple validation errors
// returned by DataVehicles.ValidateAll() if the designated constraints aren't met.
type DataVehiclesMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DataVehiclesMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DataVehiclesMultiError) AllErrors() []error { return m }

// DataVehiclesValidationError is the validation error returned by
// DataVehicles.Validate if the designated constraints aren't met.
type DataVehiclesValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataVehiclesValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataVehiclesValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataVehiclesValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataVehiclesValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataVehiclesValidationError) ErrorName() string { return "DataVehiclesValidationError" }

// Error satisfies the builtin error interface
func (e DataVehiclesValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDataVehicles.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataVehiclesValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataVehiclesValidationError{}

// Validate checks the field values on DataLicenses with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DataLicenses) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DataLicenses with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DataLicensesMultiError, or
// nil if none found.
func (m *DataLicenses) ValidateAll() error {
	return m.validate(true)
}

func (m *DataLicenses) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetLicenses()) > 200 {
		err := DataLicensesValidationError{
			field:  "Licenses",
			reason: "value must contain no more than 200 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	for idx, item := range m.GetLicenses() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DataLicensesValidationError{
						field:  fmt.Sprintf("Licenses[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DataLicensesValidationError{
						field:  fmt.Sprintf("Licenses[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DataLicensesValidationError{
					field:  fmt.Sprintf("Licenses[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return DataLicensesMultiError(errors)
	}

	return nil
}

// DataLicensesMultiError is an error wrapping multiple validation errors
// returned by DataLicenses.ValidateAll() if the designated constraints aren't met.
type DataLicensesMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DataLicensesMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DataLicensesMultiError) AllErrors() []error { return m }

// DataLicensesValidationError is the validation error returned by
// DataLicenses.Validate if the designated constraints aren't met.
type DataLicensesValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataLicensesValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataLicensesValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataLicensesValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataLicensesValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataLicensesValidationError) ErrorName() string { return "DataLicensesValidationError" }

// Error satisfies the builtin error interface
func (e DataLicensesValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDataLicenses.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataLicensesValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataLicensesValidationError{}

// Validate checks the field values on DataUserLocations with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *DataUserLocations) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DataUserLocations with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DataUserLocationsMultiError, or nil if none found.
func (m *DataUserLocations) ValidateAll() error {
	return m.validate(true)
}

func (m *DataUserLocations) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetUsers()) > 2000 {
		err := DataUserLocationsValidationError{
			field:  "Users",
			reason: "value must contain no more than 2000 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	for idx, item := range m.GetUsers() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DataUserLocationsValidationError{
						field:  fmt.Sprintf("Users[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DataUserLocationsValidationError{
						field:  fmt.Sprintf("Users[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DataUserLocationsValidationError{
					field:  fmt.Sprintf("Users[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if m.ClearAll != nil {
		// no validation rules for ClearAll
	}

	if len(errors) > 0 {
		return DataUserLocationsMultiError(errors)
	}

	return nil
}

// DataUserLocationsMultiError is an error wrapping multiple validation errors
// returned by DataUserLocations.ValidateAll() if the designated constraints
// aren't met.
type DataUserLocationsMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DataUserLocationsMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DataUserLocationsMultiError) AllErrors() []error { return m }

// DataUserLocationsValidationError is the validation error returned by
// DataUserLocations.Validate if the designated constraints aren't met.
type DataUserLocationsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DataUserLocationsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DataUserLocationsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DataUserLocationsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DataUserLocationsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DataUserLocationsValidationError) ErrorName() string {
	return "DataUserLocationsValidationError"
}

// Error satisfies the builtin error interface
func (e DataUserLocationsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDataUserLocations.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DataUserLocationsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DataUserLocationsValidationError{}

// Validate checks the field values on CitizenLocations with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *CitizenLocations) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CitizenLocations with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CitizenLocationsMultiError, or nil if none found.
func (m *CitizenLocations) ValidateAll() error {
	return m.validate(true)
}

func (m *CitizenLocations) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetIdentifier()) > 64 {
		err := CitizenLocationsValidationError{
			field:  "Identifier",
			reason: "value length must be at most 64 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetJob()) > 20 {
		err := CitizenLocationsValidationError{
			field:  "Job",
			reason: "value length must be at most 20 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetCoords() == nil {
		err := CitizenLocationsValidationError{
			field:  "Coords",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetCoords()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CitizenLocationsValidationError{
					field:  "Coords",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CitizenLocationsValidationError{
					field:  "Coords",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetCoords()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CitizenLocationsValidationError{
				field:  "Coords",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Hidden

	// no validation rules for Remove

	if len(errors) > 0 {
		return CitizenLocationsMultiError(errors)
	}

	return nil
}

// CitizenLocationsMultiError is an error wrapping multiple validation errors
// returned by CitizenLocations.ValidateAll() if the designated constraints
// aren't met.
type CitizenLocationsMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CitizenLocationsMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CitizenLocationsMultiError) AllErrors() []error { return m }

// CitizenLocationsValidationError is the validation error returned by
// CitizenLocations.Validate if the designated constraints aren't met.
type CitizenLocationsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CitizenLocationsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CitizenLocationsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CitizenLocationsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CitizenLocationsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CitizenLocationsValidationError) ErrorName() string { return "CitizenLocationsValidationError" }

// Error satisfies the builtin error interface
func (e CitizenLocationsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCitizenLocations.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CitizenLocationsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CitizenLocationsValidationError{}

// Validate checks the field values on DeleteUsers with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DeleteUsers) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteUsers with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DeleteUsersMultiError, or
// nil if none found.
func (m *DeleteUsers) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteUsers) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetUserIds()) > 100 {
		err := DeleteUsersValidationError{
			field:  "UserIds",
			reason: "value must contain no more than 100 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return DeleteUsersMultiError(errors)
	}

	return nil
}

// DeleteUsersMultiError is an error wrapping multiple validation errors
// returned by DeleteUsers.ValidateAll() if the designated constraints aren't met.
type DeleteUsersMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteUsersMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteUsersMultiError) AllErrors() []error { return m }

// DeleteUsersValidationError is the validation error returned by
// DeleteUsers.Validate if the designated constraints aren't met.
type DeleteUsersValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteUsersValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteUsersValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteUsersValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteUsersValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteUsersValidationError) ErrorName() string { return "DeleteUsersValidationError" }

// Error satisfies the builtin error interface
func (e DeleteUsersValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteUsers.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteUsersValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteUsersValidationError{}

// Validate checks the field values on DeleteVehicles with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DeleteVehicles) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteVehicles with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DeleteVehiclesMultiError,
// or nil if none found.
func (m *DeleteVehicles) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteVehicles) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetPlates()) > 100 {
		err := DeleteVehiclesValidationError{
			field:  "Plates",
			reason: "value must contain no more than 100 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return DeleteVehiclesMultiError(errors)
	}

	return nil
}

// DeleteVehiclesMultiError is an error wrapping multiple validation errors
// returned by DeleteVehicles.ValidateAll() if the designated constraints
// aren't met.
type DeleteVehiclesMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteVehiclesMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteVehiclesMultiError) AllErrors() []error { return m }

// DeleteVehiclesValidationError is the validation error returned by
// DeleteVehicles.Validate if the designated constraints aren't met.
type DeleteVehiclesValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteVehiclesValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteVehiclesValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteVehiclesValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteVehiclesValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteVehiclesValidationError) ErrorName() string { return "DeleteVehiclesValidationError" }

// Error satisfies the builtin error interface
func (e DeleteVehiclesValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteVehicles.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteVehiclesValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteVehiclesValidationError{}
